kube-derive/src/custom_resource.rs

```diff
@@ -539,6 +539,115 @@ fn generate_hasspec(spec_ident: &Ident, root_ident: &Ident, kube_core: &Path) ->
     }
 }
 
+#[derive(FromAttributes)]
+#[darling(attributes(validated))]
+struct CELAttr {
+    rule: String,
+    message: Option<String>,
+}
+
+pub(crate) fn cel_validation(_: TokenStream, input: TokenStream) -> TokenStream {
+    let mut ast: DeriveInput = match syn::parse2(input) {
+        Err(err) => return err.to_compile_error(),
+        Ok(di) => di,
+    };
+
+    if !ast.attrs.iter().any(|attr| attr.path().is_ident("derive")) {
+        return syn::Error::new(
+            ast.ident.span(),
+            r#"#[cel_validation] macro should be placed before the #[derive(JsonSchema)] macro"#,
+        )
+        .to_compile_error();
+    }
+
+    let struct_name = ast.ident.to_string() + "Validation";
+    let anchor = Ident::new(&struct_name, Span::call_site());
+
+    let mut validations: Vec<TokenStream> = vec![];
+
+    let struct_data = match ast.data {
+        syn::Data::Struct(ref mut struct_data) => struct_data,
+        _ => {
+            return syn::Error::new(
+                ast.ident.span(),
+                r#"#[cel_validation] has to be used with structs"#,
+            )
+            .to_compile_error()
+        }
+    };
+
+    if let syn::Fields::Named(fields) = &mut struct_data.fields {
+        for field in &mut fields.named {
+            let mut rules = vec![];
+            for attr in field
+                .attrs
+                .iter()
+                .filter(|attr| attr.path().is_ident("validated"))
+            {
+                let CELAttr { rule, message } = match CELAttr::from_attributes(&vec![attr.clone()]) {
+                    Ok(cel) => cel,
+                    Err(e) => return e.with_span(&attr.meta).write_errors(),
+                };
+                let message = if let Some(message) = message {
+                    quote! { "message": #message }
+                } else {
+                    quote! {}
+                };
+                rules.push(quote! {{
+                    "rule": #rule,
+                    #message
+                },});
+            }
+
+            if rules.is_empty() {
+                continue;
+            }
+
+            let validation_method_name = field.ident.as_ref().map(|i| i.to_string()).unwrap_or_default();
+            let name = Ident::new(&validation_method_name, Span::call_site());
+            let field_type = &field.ty;
```

Comment:
This stuff here could do with some comments. You're implementing a method fith a fixed name. Could this ave clashing issues? Can we parametrise such a function from kube-core instead?

Reply:
As long as there is no structure <name>Validation in the scope, there should not be a clash, but makes sense to add an override for such occasion 👍🏻

Reply:
Should be added now - `#[cel_validation(struct_name = "FooSpecValidation")]`

Reply:
i guess i am still confused about the temporary struct you are making. it seems to me that the #[cel_validation(struct_name = X struct X is there for you to implement a method, but that method feels like something we can define cleanly inside kube-core, and invoke from kube-derive.